Implemented JWT authentication support.

ShastkivRuslan opened the pull request:
Implemented JWT authentication support:
- Added `/api/auth/login` endpoint for user authentication.
- Created `AuthenticationService`, `JwtAuthenticationFilter`, and `JwtUtil` classes.
- Added validation for data in the new DTO classes.
- Configured JWT properties in `application.properties`.
- Integrated JWT filter into the `SecurityFilterChain` bean for processing client requests.
- Added custom exception handlers for AuthenticationException and BadCredentialsException
[jwt](https://github.com/ShastkivRuslan/online-book-shop/tree/jwt)

ShastkivRuslan committed:
-added return type Page<> instead List in some endpoints;

ShastkivRuslan committed:
Implemented JWT authentication support:

- Added `/api/auth/login` endpoint for user authentication.
- Created `AuthenticationService`, `JwtAuthenticationFilter`, and `JwtUtil` classes.
- Added validation for data in the new DTO classes.
- Configured JWT properties in `application.properties`.
- Integrated JWT filter into the `SecurityFilterChain` bean for processing client requests.
- Added custom exception handlers for AuthenticationException and BadCredentialsException

ShastkivRuslan committed:
-refactoring -> done;

ShastkivRuslan committed:
-refactoring_v2 -> done;

Elena-Bruyako commented on src/main/java/ruslan/shastkiv/bookstore/controller/AuthenticationController.java:
```diff
@@ -29,4 +33,9 @@ public UserDto register(@RequestBody @Valid UserRegistrationRequestDto registrat
         return userService.register(registrationRequestDto);
     }
+
+    @PostMapping("/login")
```
add Swagger annotation

Elena-Bruyako commented on src/main/java/ruslan/shastkiv/bookstore/dto/user/UserLoginRequestDto.java:
```diff
@@ -0,0 +1,15 @@
+
+public record UserLoginRequestDto(
+        @Email
+        @NotEmpty
```
```suggestion
        @NotBlank
```

Elena-Bruyako commented on src/main/java/ruslan/shastkiv/bookstore/dto/user/UserLoginRequestDto.java:
```diff
@@ -0,0 +1,15 @@
+        @NotEmpty
+        @Size(min = 9, max = 50)
+        String email,
+        @NotEmpty
```
```suggestion
        @NotBlank
```

Elena-Bruyako commented on src/main/java/ruslan/shastkiv/bookstore/security/JwtAuthenticationFilter.java:
```diff
@@ -0,0 +1,49 @@
+
+    private String getToken(HttpServletRequest request) {
+        String bearerToken = request.getHeader("Authorization");
+        if (StringUtils.hasText(bearerToken) && bearerToken.startsWith("Bearer ")) {
```
create constant for `"Bearer "`

Elena-Bruyako commented on src/main/java/ruslan/shastkiv/bookstore/security/JwtAuthenticationFilter.java:
```diff
@@ -0,0 +1,49 @@
+    private String getToken(HttpServletRequest request) {
+        String bearerToken = request.getHeader("Authorization");
+        if (StringUtils.hasText(bearerToken) && bearerToken.startsWith("Bearer ")) {
+            return bearerToken.substring(7);
```
use constant.length() instead of 7

Elena-Bruyako commented on src/main/java/ruslan/shastkiv/bookstore/security/JwtAuthenticationFilter.java:
```diff
@@ -0,0 +1,49 @@
+    }
+
+    private String getToken(HttpServletRequest request) {
+        String bearerToken = request.getHeader("Authorization");
```
```suggestion
        String bearerToken = request.getHeader(HttpHeaders.AUTHORIZATION);
```

Elena-Bruyako commented on src/main/java/ruslan/shastkiv/bookstore/exception/GlobalExceptionHandler.java:
```diff
@@ -44,4 +47,26 @@ public ResponseEntity<Map<String, String>> handleRegistrationException(
+        return new ResponseEntity<>(errorResponse, HttpStatus.UNAUTHORIZED);
+    }
+
+    @ExceptionHandler(BadCredentialsException.class)
```
where do you get this exception?

Elena-Bruyako commented on src/main/java/ruslan/shastkiv/bookstore/exception/GlobalExceptionHandler.java:
```diff
@@ -44,4 +47,26 @@ public ResponseEntity<Map<String, String>> handleRegistrationException(
+        return new ResponseEntity<>(errorResponse, HttpStatus.UNAUTHORIZED);
+    }
+
+    @ExceptionHandler(JwtException.class)
```
JwtException won't be handled by GlobalExceptionHandler because it appears in filter, so you can avoid to handle it or find the way to handle correctly

Elena-Bruyako reviewed (changes requested):
see comments

ShastkivRuslan committed:
Refactor Authentication and Exception Handling based on mentor feedback

- Added Swagger annotation to the `/login` endpoint in AuthenticationController
- Updated UserLoginRequestDto:
  - Added `@NotBlank` annotation to `email` and `password` fields for stricter validation
- Refactored JwtAuthenticationFilter:
  - Replaced `"Bearer "` with a constant for readability and reusability
  - Used `constant.length()` instead of hardcoded value `7` for substring extraction
  - Replaced `request.getHeader("Authorization")` with `HttpHeaders.AUTHORIZATION` for better readability
